aten/src/ATen/native/sparse/SparseBinaryOpIntersectionCommon.h

```diff
@@ -0,0 +1,388 @@
+#pragma once
+
+#define TORCH_ASSERT_ONLY_METHOD_OPERATORS
+#include <ATen/Tensor.h>
+#include <ATen/native/TensorIterator.h>
+#include <ATen/Dispatch.h>
+#include <ATen/native/sparse/Macros.h>
+#include <ATen/ExpandUtils.h>
+#include <ATen/SparseTensorUtils.h>
+#include <iostream>
+
+#ifndef AT_PER_OPERATOR_HEADERS
+#include <ATen/Functions.h>
+#include <ATen/NativeFunctions.h>
+#else
+#include <ATen/ops/arange.h>
+#include <ATen/ops/empty.h>
+#include <ATen/ops/_sparse_coo_tensor_with_dims_and_tensors.h>
+#include <ATen/ops/from_blob.h>
+#include <ATen/ops/result_type.h>
+#endif
+
+#ifdef GPUCC
+#define NAME "sparse_binary_op_intersection_cuda"
+#else
+#define NAME "sparse_binary_op_intersection_cuda"
+#endif
+
+#define BINARY_OP_API static constexpr FUNCAPI INLINE
+
+namespace at {
+namespace native {
+
+namespace {
+
+using at::sparse::get_sparse_impl;
+
+// ForwardIt: only legacy random access iterator is supported.
+template<class ForwardIt, class T, bool is_lower = true>
+static FUNCAPI INLINE
+ForwardIt find_bound(ForwardIt RESTRICT first, ForwardIt RESTRICT last, const T& value) {
+    ForwardIt RESTRICT it;
+    typename std::iterator_traits<ForwardIt>::difference_type count, step;
+    // NOTE: std::distance(first, last) compiles but produces wrong results on CUDA,
+    // so only legacy random access iterators are safe in this code.
+    count = last - first;
+
+    while (count > 0) {
+      it = first;
+      step = count / 2;
+      // avoiding std::advance(it, step),
+      // although it does work unlike std::distance on CUDA.
+      it += step;
+      if (is_lower ? *it < value : value >= *it) {
+        first = ++it;
+        count -= step + 1;
+      }
+      else {
+        count = step;
+      }
+    }
+    return first;
+}
+
+template <template <typename func_t> class kernel_t>
+struct KernelLauncher {
+  template <typename func_t>
+  static void launch(TensorIteratorBase& iter, const func_t& f) {
+    kernel_t<func_t>::launch(iter, f);
+  }
+};
+
+template <
+  template <typename func_t> class kernel_t,
+  typename binary_op_t,
+  typename hash_t = int64_t,
+  typename offset_t = int64_t>
+Tensor _sparse_binary_op_intersection_kernel_impl(
+    Tensor& res,
+    const Tensor& x_,
+    const Tensor& y_,
+    const std::vector<int64_t> broadcasted_shape,
```

Comment:
For my own edification. Is there a reason why `std::vector<>` is needed or preferred over `DimVector`/`IntArrayRef`? It is not obvious to me if there is.

Reply:
It is being wrapped into a tensor with from_blob which takes in a pointer. DimVector does not survive untill the point I need it's data.